stl/inc/memory

```diff
@@ -2406,26 +2427,273 @@ bool atomic_compare_exchange_weak_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty
+    using _Base = _Atomic_ptr_base<_Ty>;
+
+public:
+    using value_type = shared_ptr<_Ty>;
+
+    static constexpr bool is_always_lock_free = false;
+
+    _NODISCARD bool is_lock_free() const noexcept {
+        return false;
+    }
+
+    void store(shared_ptr<_Ty> _Value, const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_store_memory_order(_Order);
+        const auto _Rep = this->_Lock_and_load();
+        _STD swap(this->_Ptr, _Value._Ptr);
+        this->_Store_and_unlock(_Value._Rep);
+        _Value._Rep = _Rep;
+    }
+
+    _NODISCARD shared_ptr<_Ty> load(const memory_order _Order = memory_order::seq_cst) const noexcept {
+        _Check_load_memory_order(_Order);
+        shared_ptr<_Ty> _Result;
+        const auto _Rep = this->_Lock_and_load();
+        _Result._Ptr    = this->_Ptr;
+        _Result._Rep    = _Rep;
+        _Result._Incref();
+        this->_Store_and_unlock(_Rep);
+        return _Result;
+    }
+
+    operator shared_ptr<_Ty>() const noexcept {
+        return load();
+    }
+
+    _NODISCARD shared_ptr<_Ty> exchange(
+        shared_ptr<_Ty> _Value, const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_memory_order(_Order);
+        shared_ptr<_Ty> _Result;
+        _Result._Rep = this->_Lock_and_load();
+        _Result._Ptr = this->_Ptr;
+        this->_Ptr   = _Value._Ptr;
+        this->_Store_and_unlock(_Value._Rep);
+        _Value._Ptr = nullptr; // ownership of this ref has been given to this, silence decrement
+        _Value._Rep = nullptr;
+        return _Result;
+    }
+
+    bool compare_exchange_weak(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired, const memory_order _Success,
+        const memory_order _Failure) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Combine_cas_memory_orders(_Success, _Failure));
+    }
+
+    bool compare_exchange_strong(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired, const memory_order _Success,
+        const memory_order _Failure) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Combine_cas_memory_orders(_Success, _Failure));
+    }
+
+    bool compare_exchange_weak(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired,
+        const memory_order _Order = memory_order::seq_cst) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Order);
+    }
+
+    bool compare_exchange_strong(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired,
+        const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_memory_order(_Order);
+        auto _Rep = this->_Lock_and_load();
+        if (this->_Ptr == _Expected._Ptr && _Rep == _Expected._Rep) {
+            _STD swap(this->_Ptr, _Desired._Ptr);
+            _STD swap(_Rep, _Desired._Rep);
+            this->_Store_and_unlock(_Rep);
+            return true;
+        } else {
+            _Ref_count_base* _ExpectedRep = _Expected._Rep;

```

Comment:
@StephanTLavavej Can you double check that this load op is reasonable? (and for the weak_ptr version?) It looks correct to me but this is the only part I consider "tricky" and you are the "shared_ptr czar".